store/postgres/src/store.rs

```diff
@@ -997,40 +999,86 @@ impl StoreTrait for Store {
+                    Some(inner) => inner,
+                    None => return Ok(None),
+                };
 
-            let by_causality_region = entities
-                .into_iter()
-                .map(|e| {
-                    let causality_region = e.id()?;
-                    let digest = match e.get("digest") {
-                        None => Err(format_err!("Entity is missing a digest attribute")),
-                        Some(Value::String(s)) => Ok(s.to_owned()),
-                        _ => Err(format_err!("Entity has non-string digest attribute")),
-                    }?;
+                cancel.check_cancel()?;
+
+                // FIXME: (Determinism)
+                // There is no guarantee that we are able to look up the block number
+                // even if we have indexed beyond it, because the cache is sparse and
+                // only populated by blocks with triggers.
+                //
+                // This is labeled as a determinism bug because the PoI needs to be
+                // submitted to the network reliably. Strictly speaking, that's not
+                // indeterminism to miss an opportunity to claim a reward, but it's very
+                // similar to most determinism bugs in that money is on the line.
+                let block_number = self_inner
+                    .block_number(&subgraph_id_inner, block_hash)
+                    .map_err(|e| CancelableError::Error(e.into()))?;
+                let block_number = match block_number {
+                    Some(n) => n.try_into().unwrap(),
+                    None => return Ok(None),
+                };
+                cancel.check_cancel()?;
+
+                // FIXME: (Determinism)
+                // It is vital to ensure that the block hash given in the query
+                // is a parent of the latest block indexed for the subgraph.
+                // Unfortunately the machinery needed to do this is not yet in place.
+                // The best we can do right now is just to make sure that the block number
+                // is high enough.
+                if latest_block_ptr.number < block_number {
+                    return Ok(None);
+                }
 
-                    let poi = ProofOfIndexingDigest(digest);
+                let entities = conn
+                    .query(
+                        &logger,
+                        EntityCollection::All(vec![POI_OBJECT.to_owned()]),
+                        None,
+                        None,
+                        EntityRange {
+                            first: None,
+                            skip: 0,
+                        },
+                        block_number.try_into().unwrap(),
+                    )
+                    .map_err(Error::from)?;
 
-                    Ok((causality_region, poi))
-                })
-                .collect::<Result<HashMap<_, _>, Error>>()?;
+                cancel.check_cancel()?;
```

Comment:
We could already leave the transaction scope here, to release the lock on the table earlier, since we no longer need access to the database. The gain would be insignificant now, but with many causality regions, putting the final PoI together could take some time.

Reply:
Finishing the hash is practically instant on the database time scale, but I'll fix this anyway.